### PR TITLE
Implement entity materialization interception

### LINKED ISSUE #15911: Implement ObjectMaterialized event
Splitting this out as a separate issue as required on #626 so we can track it independently of other life-cycle hooks.

The `ObjectMaterialized` event will fire after an entity instance has been created and all it's non-navigation properties have been set.

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #626
Fixes #15911

Introduces a new `IMaterializationInterceptor` singleton interceptor that allows:
- Interception before any entity instance has been created, allowing a customized instance to be created (if desired), thereby suppressing of normal EF instance creation.
- Interception after the instance has been created, but before property values have been set. The instance can be replaced with a new instance (if desired), without preventing EF from setting property values.
- Interception before property values have been set, allowing custom setting of the values and/or suppression of setting the values by EF (if desired).
- Interception after property values have been set, allowing the instance to be changed (if desired.)

Access to property values, including shadow and service properties is provided at each point.

If no singleton materialization interceptors are configured, then the materialization delegate is the same as before, meaning any perf impact only applies if interception is used.

### CODE REVIEW

File: src/EFCore/Diagnostics/IMaterializationInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    /// <param name="instance">The entity instance for which property values will be set.</param>
+    /// <param name="result">
+    ///     Represents the current result if one exists.
+    ///     This value will have <see cref="InterceptionResult.IsSuppressed" /> set to <see langword="true" /> if some previous
+    ///     interceptor suppressed execution by calling <see cref="InterceptionResult.Suppress" />.
+    ///     This value is typically used as the return value for the implementation of this method.
+    /// </param>
+    /// <returns>
+    ///     If <see cref="InterceptionResult.IsSuppressed" /> is false, the EF will continue as normal.
+    ///     If <see cref="InterceptionResult.IsSuppressed" /> is true, then EF will not set any property values.
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/IMaterializationInterceptor.cs:66]:
```suggestion
    ///     If <see cref="InterceptionResult.IsSuppressed" /> is <see langword="false" />, the EF will continue as normal.
    ///     If <see cref="InterceptionResult.IsSuppressed" /> is <see langword="true" />, then EF will not set any property values.
```

File: src/EFCore/Diagnostics/MaterializationInterceptionData.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    /// <summary>
+    ///     Creates a dictionary containing a snapshot of all property values for the instance being materialized.
+    /// </summary>
+    /// <remarks>
+    ///     Note that this method causes a new dictionary to be created, and copies all values into that dictionary. This can
+    ///     be less efficient than getting values individually, especially if the non-boxing generic overloads
+    ///     <see cref="GetPropertyValue{T}(string)" /> or <see cref="GetPropertyValue{T}(IPropertyBase)" />can be used.
+    /// </remarks>
+    /// <returns>The values of properties for the entity instance.</returns>
+    public IReadOnlyDictionary<IPropertyBase, object?> CreateValuesDictionary()
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/MaterializationInterceptionData.cs:119]: Why is this needed?

File: src/EFCore/Diagnostics/MaterializationInterceptionData.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+///     A parameter object passed to <see cref="IMaterializationInterceptor" /> methods containing data about the instance
+///     being materialized.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-diagnostics">Logging, events, and diagnostics</see> for more information and examples.
+/// </remarks>
+public readonly struct MaterializationInterceptionData
+{
+    private readonly MaterializationContext _materializationContext;
+    private readonly IDictionary<IPropertyBase, (object TypedAccessor, Func<MaterializationContext, object?> Accessor)> _valueAccessor;
```
Comment (AndriySvyryd) [inline src/EFCore/Diagnostics/MaterializationInterceptionData.cs:18]: You can key the dictionary by name to avoid property lookups

Comment (ajcvickers) [reply src/EFCore/Diagnostics/MaterializationInterceptionData.cs:18]: Still have to do a property lookup for validation, and keying by object reference should be faster for cases where the property is already known, such as when driving this from metadata.

File: src/EFCore/Diagnostics/MaterializationInterceptionData.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore/Diagnostics/MaterializationInterceptionData.cs:119]: I thought it would be helpful for people who wanted an easy-to-use snapshot for values. They could, of course, write it themselves.

Comment (ajcvickers) [reply src/EFCore/Diagnostics/MaterializationInterceptionData.cs:119]: Happy to remove it. We'll see how long it is before someone asks how to do it. :-)
